File: SafeCityAPI/src/main/java/com/ph/service/IncidentServiceImpl.java
package com.ph.service;

import com.ph.dao.IncidentDao;
import com.ph.form.IncidentForm;
import com.ph.model.Incident;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.geo.GeoJsonPoint;
import org.springframework.stereotype.Service;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;

/**
 * Created by leon on 25/05/2017.
 */
@Service("incidentService")
public class IncidentServiceImpl implements IncidentService {

    @Autowired
    private IncidentDao incidentDao;

    public Incident convertFormToIncident(IncidentForm incidentForm) {
        return new Incident(
                new GeoJsonPoint(Double.parseDouble(incidentForm.getLng()), Double.parseDouble(incidentForm.getLat())),
                incidentForm.getTitle(),
                incidentForm.getDescription(),
                incidentForm.getIncident(),
                getIncidentLevel(incidentForm.getIncident()),
                getDateFormat(incidentForm.getDate())
        );
    }

    @Override
    public List<Incident> findNearIncidents(double lat, double lng) {
        return incidentDao.getNearIncidents(lat, lng);
    }

    @Override
    public void save(Incident incident) {
        incidentDao.save(incident);
    }

    private int getIncidentLevel(String incident) {
        switch (incident) {
            case "Fight":
                return 1;
            case "Big Fight":
                return 2;
            case "Theft":
                return 3;
            case "Establishment Theft":
                return 4;
            case "House Theft":
                return 5;
            case "Shooting":
                return 6;
            case "Murder":
                return 7;
            case "Drugs":
                return 8;
            case "Police Arrest":
                return 5;
            case "kidnapping":
                return 5;
            default:
                return 3;
        }

    }

    private Date getDateFormat(String date) {
        DateFormat format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm");
        try {
            return format.parse(date);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return new Date();
    }

}

File: SafeCityAPI/src/main/java/com/ph/service/IncidentServiceImpl.kt
package com.ph.service

import com.ph.dao.IncidentDao
import com.ph.dto.TimeIncidentCount
import com.ph.form.IncidentForm
import com.ph.model.Incident
import org.springframework.data.mongodb.core.geo.GeoJsonPoint
import org.springframework.stereotype.Service
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

@Service("incidentService")
class IncidentServiceImpl(private val incidentDao: IncidentDao) : IncidentService {

    override fun convertFormToIncident(incidentForm: IncidentForm): Incident
            = Incident(
            GeoJsonPoint(java.lang.Double.parseDouble(incidentForm.lng), java.lang.Double.parseDouble(incidentForm.lat)),
            incidentForm.title,
            incidentForm.description,
            incidentForm.incident,
            getIncidentLevel(incidentForm.incident),
            getDateFormat(incidentForm.date)
    )


    override fun findNearIncidents(lat: Double, lng: Double): kotlin.collections.List<Incident> = incidentDao.getNearIncidents(lat, lng)


    override fun save(incident: Incident) {
        incidentDao.save(incident)
    }

    override fun getStatisticsByZone(lat: Double, lng: Double): MutableList<TimeIncidentCount> {
        return incidentDao.getStatistics(lat, lng)
    }



    private fun getIncidentLevel(incident: String): Int =
            when (incident) {
                "Fight" -> 1
                "Big Fight" -> 2
                "Theft" -> 3
                "Establishment Theft" -> 4
                "House Theft" -> 5
                "Shooting" -> 6
                "Murder" -> 7
                "Drugs" -> 8
                "Police Arrest" -> 5
                "kidnapping" -> 5
                "Rape" -> 8
                else -> 3
            }

    private fun getDateFormat(date: String): Date {
        val format = SimpleDateFormat("yyyy-MM-dd'T'HH:mm")
        try {
            return format.parse(date)
        } catch (e: ParseException) {
            e.printStackTrace()
        }
        return Date()
    }
}